File: day2/part1.ts
import fs from "fs";
import path from "path";

const input = fs.readFileSync(path.join(__dirname, "input.txt"), "utf8");

const cubeMap = new Map([
  ["red", 12],
  ["green", 13],
  ["blue", 14],
]);

const sum = input.split("\n").reduce((acc: number, line: string) => {
  const [identifier, data] = line.split(":");

  const id = parseInt(identifier.slice(5, identifier.length));

  const sets = data.split(";").map((set) => set.trim());

  const invalid = sets.some((set) => {
    const cubes = set.split(",").map((cube) => cube.trim());
    return cubes.some((cube) => {
      const [count, colour] = cube.split(" ");
      return parseInt(count) > cubeMap.get(colour)!;
    });
  });

  if (invalid) {
    return acc;
  }

  return acc + id;
}, 0);

console.log(sum);

export {};

File: day02/src/bin/part1.rs
use std::collections::HashMap;

fn main() {
    let input = include_str!("./input.txt");
    let output = process(input);
    dbg!(output);
}

fn process(input: &str) -> i32 {
    let cube_map = HashMap::from([
        ("red", 12),
        ("green", 13),
        ("blue", 14),
    ]);

    let mut sum = 0;
    for game in input.lines() {
        let (identifier, data) = game.split_once(": ").unwrap();

        let id = &identifier[5..identifier.len()].parse::<i32>().unwrap();
        let sets: Vec<&str> = data.split("; ").collect();

        if !sets.iter().any(|set| {
            let cubes: Vec<&str> = set.split(", ").collect();
            return cubes.iter().any(|cube| {
                let (size, colour) = cube.split_once(" ").unwrap();
                return size.parse::<i32>().unwrap() > *cube_map.get(colour).unwrap();
            });
        }) {
            sum += id;
        }
    }

    sum
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let input = include_str!("test.txt");
        let output = process(input);
        assert_eq!(output, 8);
    }
}
